src/input-states.js

```diff
@@ -0,0 +1,256 @@
+/**
+ * input-states.js
+ * 
+ * Manages RDF pipeline input states for a VNI
+ * TODO: Add optimization for input states as a sorted array
+ */
+
+var _ = require('underscore');
+
+var InputType = {
+    IIP: 1,
+    PACKET: 2
+};
+var ALL_PORTINFO_KEYS = [ 'inportName', 'sourceNodeName', 'sourcePortName' ];
+
+var inputStates = [];
+
+/**
+ * Get the index to the state in the input states array if it exists already
+ * 
+ * @param portInfo an object that describes the port(s) associated with the data
+ *
+ * @return an integer index into the inputStates array if a state is found or -1 if no state was found.
+ */
+function getStateIndex( portInfo ) { 
+
+    var portSourceNodeName = portInfo.sourceNodeName || '';
+    var portSourcePortName = portInfo.sourcePortName || '';
+
+    return _.findIndex( inputStates, function(inputState) { 
+        
+        var stateSourceNodeName = inputState.sourceNodeName || '';
+        var stateSourcePortName = inputState.sourcePortName || '';
+
+        return ( portInfo.inportName === inputState.inportName &&
+                 portSourceNodeName === stateSourceNodeName &&
+                 portSourcePortName === stateSourcePortName );
+    });
+}
+
+
+function insertAtIndex( portInfo ) { 
+
+    var portSourceNodeName = portInfo.sourceNodeName || '';
+    var portSourcePortName = portInfo.sourcePortName || '';
+
+    for ( var i=0, max=inputStates.length; i < max; i++ ) { 
+  
+        var inputState = inputStates[i];
+        var stateSourceNodeName = inputState.sourceNodeName || '';
+        var stateSourcePortName = inputState.sourcePortName || '';
+
+        if ( portInfo.inportName <= inputState.inportName &&
+             portSourceNodeName <= stateSourceNodeName &&
+             portSourcePortName <= stateSourcePortName ) {
+             
+             return i;
+        }
```

Comment:
I think it would be good to insert them in sorted order, but I don't think the above 'if' statement does that.  I suggest something like:
<pre>
// Insert in sorted order
if (portInfo.inportName < inputState.inportName) { return i; }
if (portInfo.inportName > inputState.inportName) { continue; }
// portInfo.inportName === inputState.inportName
if (portSourceNodeName < stateSourceNodeName) { return i; }
if (portSourceNodeName > stateSourceNodeName) { continue; }
// portSourceNodeName === stateSourceNodeName
if (portSourcePortName < stateSourcePortName) { return i; }
if (portSourcePortName > stateSourcePortName) { continue; }
// portSourcePortName === stateSourcePortName
break;
</pre>